### PR TITLE
Query for ctx

### PR DESCRIPTION
**Pull request by akolomietc**

### CODE REVIEW

File: database/database.go [hunk 1]
```diff
@@ -74,6 +105,7 @@ type Database struct {
 	isReady   int32
 	errors    *errorsWatch
 	WarnChan  chan []error
+	before    []BeforeQueryFunc
```
Comment (avdva) [inline database/database.go]: it seems, that there's need to protect this array with a mutex.

File: database/database.go [hunk 2]
```diff
@@ -61,6 +62,36 @@ type Config struct {
...
 
+type BeforeQueryFunc func(ctx context.Context, query string, args ...interface{}) context.Context
+
+func (extDb *Database) SetBeforeQuery(f BeforeQueryFunc) {
+	extDb.before = append(extDb.before, f)
+}
+
+func (extDb *Database) ExecWithContext(ctx context.Context, query string, args ...interface{}) (sql.Result, error) {
+	for _, f := range extDb.before {
+		ctx = f(ctx, query, args)
```
Comment (avdva) [inline database/database.go]: I'm not sure, that it is a mistake, but anyway

consider the following code:

```package main

import (
	"fmt"
)

type ftype func(args ...interface{})

func exec(f ftype, args ...interface{}) {
	f(args)
}

func exec2(f ftype, args ...interface{}) {
	f(args...)
}

func main() {
	f := func(args ...interface{}) {
		for _, arg := range args {
			fmt.Printf("%+v\n", arg)
		}
	}
	exec(f, "a", "b")
	exec2(f, "a", "b")
}
```
In `exec2`, unlike `exec` the arguments are passed via `...`, and as a result you have several arguments, not a slice.
I'm not sure, that slices don't work in your case, but, if you still want different arguments, not a slice, you should use `...`

File: database/database.go [hunk 3]
```diff
@@ -61,6 +62,36 @@ type Config struct {
 	SSLMode        string `yaml:"ssl_mode"`
 }
 
+type BeforeQueryFunc func(ctx context.Context, query string, args ...interface{}) context.Context
+
+func (extDb *Database) SetBeforeQuery(f BeforeQueryFunc) {
```
Comment (avdva) [inline database/database.go]: The name doesn't reflect what the function does. Either rename the function, or change its code in a way, when in sets the before func.

File: database/database.go [hunk 4]
```diff
@@ -61,6 +62,36 @@ type Config struct {
 	SSLMode        string `yaml:"ssl_mode"`
 }
 
+type BeforeQueryFunc func(ctx context.Context, query string, args ...interface{}) context.Context
```
Comment (avdva) [inline database/database.go]: Comments for all exported names?

**Review by avdva — APPROVED** [review]

File: database/database_test.go [hunk 5]
```diff
@@ -64,3 +66,27 @@ func doTest(t *testing.T, db *Database) {
 	require.Nil(t, row)
 	require.Equal(t, sql.ErrNoRows, err)
 }
+
+func doTestCtx(t *testing.T, db *Database) {
+	_, err := db.Exec("delete from t1")
+	require.NoError(t, err)
+
+	_, err = db.ExecWithContext(context.TODO(), `insert into t1(id, v) values($1,$2)`, 2, 3)
+	require.Nil(t, err)
```
Comment (avdva) [inline database/database_test.go]: why not `NoError`, as above?

File: tracing/database.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	"context"
+	"database/sql"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+	"github.com/opentracing/opentracing-go/log"
+	"github.com/space307/go-utils/database"
+)
+
+type Database struct {
```
Comment (avdva) [inline tracing/database.go:15]: comments?

**Review by s0rg — APPROVED** [review]

File: tracing/database.go (hunk 6 above)
Comment (akolomietc) [reply tracing/database.go:15]: Yes, will do

File: tracing/database.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	ExtDB *database.Database
```
Comment (avdva) [inline tracing/database.go]: Is there need for it to be public?

Comment (akolomietc) [reply tracing/database.go]: Yes, if I need use simple database.Database object

Comment (b13f) [reply tracing/database.go]: Togda luche ispolzovat' getter

Comment (akolomietc) [reply tracing/database.go]: I do not understand you @b13f

Comment (b13f) [reply tracing/database.go]: @pilotuser TY DOLZHEN ISPOLZOVAT GETTER

Comment (avdva) [reply tracing/database.go]: In my mind, we use this wrapper for traced execution only.
If we don't need tracing, and just want  database.Database, we should have it, not retrieve from this wrapper.

File: tracing/database.go [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+// Init creates a storage object based on a given config.
+func Init(driver string, config *database.Config) (*Database, error) {
+	extDB, err := database.Init(driver, config)
+	return &Database{extDB: extDB}, err
+}
+
+// GetExtDatabase getter for database.Database object
+func (d *Database) GetExtDatabase() *database.Database {
```
Comment (avdva) [inline tracing/database.go:26]: ok, let it be.

**Review by avdva — APPROVED** [review]

File: tracing/database.go (hunk 7 above)
Comment (akolomietc) [reply tracing/database.go]: ok, I added a getter.

**Pull request merged**